### PR TITLE
Feature/backend/enhance pi v0

### PR DESCRIPTION
**Pull request by gdchamal**
Enhance privacy features extraction and PI compute for an inbound message.

- Processing is outside of message parsing logic and related code is moved into caliopen_pi package for clarification.

- Spam qualification use more headers than previous method, may cover most antispam solution.

- Fix contact_lookup insert to get contacts informations  during PI compute

### CODE REVIEW

File: src/backend/main/py.pi/caliopen_pi/features/ingress_path.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    # Second step: qualify path if internal and try to find the ingress one
+    ingress = None
+    internal_hops = 0
+    for path in paths:
+        is_internal = False
+        for internal in internal_domains:
+            if internal in path[0]:
+                is_internal = True
+            else:
+                if internal in path[1] and internal not in path[0]:
```
Comment (sapiens-sapide) [inline src/backend/main/py.pi/caliopen_pi/features/ingress_path.py]: `internal not in path[0]` not needed line 64, is it ?

File: src/backend/main/py.pi/caliopen_pi/features/mail.py [hunk 2]
```diff
@@ -184,3 +185,9 @@ def compute_pi(self, participants, features):
                             'context': sum(pi_cx.values()),
                             'comportment': sum(pi_co.values()),
                             'version': 0})
+
+    def process(self, message, participants):
+        """Process the message for privacy features and PI compute."""
```
Comment (sapiens-sapide) [inline src/backend/main/py.pi/caliopen_pi/features/mail.py]: Could we have a more verbose func definition ? : 

```
""" Compute privacy features and PI from self and put them into message

   :param message: a NewInboundMessage(). Will be updated with PI and privacy features.
   :param participants: an array of Participants()
"""
```

File: src/backend/main/py.pi/caliopen_pi/features/types.py [hunk 3]
```diff
@@ -81,4 +81,8 @@ def marshall_features(features):
 
 def unmarshall_features(features):
     """Unmarshall a dict of features suitable for storage."""
-    return dict((name, str(value)) for name, value in features.items())
+    res = {}
+    for k, v in features.items():
+        if not (v == '' or v is None):
```
Comment (sapiens-sapide) [inline src/backend/main/py.pi/caliopen_pi/features/types.py]: `if v:` is enough ;-)

Comment (gdchamal) [inline src/backend/main/py.pi/caliopen_pi/features/types.py]: Nope I want 0 to be serialized but not None and ''

File: src/backend/main/py.pi/caliopen_pi/features/ingress_path.py (hunk 1 above)
Comment (gdchamal) [inline src/backend/main/py.pi/caliopen_pi/features/ingress_path.py]: yes you're right

File: src/backend/main/py.pi/caliopen_pi/features/mail.py [hunk 4]
```diff
@@ -189,7 +189,14 @@ def _compute_pi(self, participants, features):
                             'version': 0})
 
     def process(self, message, participants):
-        """Process the message for privacy features and PI compute."""
+        """
+        Process the message for privacy features and PI compute.
+
+        :param message: a message parameter that will be updated with PI
+        :ptype message: NewMessage
```
Comment (sapiens-sapide) [inline src/backend/main/py.pi/caliopen_pi/features/mail.py]: `:ptype message: NewInboundMessage`

Comment (gdchamal) [inline src/backend/main/py.pi/caliopen_pi/features/mail.py]: I'll fix in another branch this documentation.I didn't see appearing this NewInboundMessage parameter

**Pull request merged**